Formatter doesn't have access to filename

#2942 added `formatter` option, which is very useful.
I used it prettier but quickly discovered that I have conflicting formatting between CI and my editor.

This happens because prettier change formatting based on filename, e.g. `package.json` formatted with  `json-stringify` but other `*.json` files formatted with `json` parser.
https://github.com/prettier/prettier/blob/2316e2fecd171335fc00f3ec97b7e46cc8964153/src/language-js/index.js#L81-L94

To solve this use case, prettier provides [--stdin-filepath](https://prettier.io/docs/en/cli.html#--stdin-filepath), but I can't use it in helix since there is no way to pass the filename into `fomatter` option.

Having a placeholder of some kind would be great, e.g. `$filename`.

It gives me the error:  2024-07-19T19:10:00.041 helix_view::document [ERROR] Formatter error: error: no files provided
And if i provide no filename argument i get the output of stylua asking for the argument to --stdin-filepath

--- a/book/src/languages.md
+++ b/book/src/languages.md
@@ -66,7 +66,7 @@ These configuration keys are available:
 | `indent`              | The indent to use. Has sub keys `unit` (the text inserted into the document when indenting; usually set to N spaces or `"\t"` for tabs) and `tab-width` (the number of spaces rendered for a tab) |
 | `language-servers`    | The Language Servers used for this language. See below for more information in the section [Configuring Language Servers for a language](#configuring-language-servers-for-a-language)   |
 | `grammar`             | The tree-sitter grammar to use (defaults to the value of `name`) |
-| `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout |
+| `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout. The filename of the current buffer can be passed as argument by using the `%{buffer_name}` expansion variable. See below for more information in the [Configuring the formatter command](#configuring-the-formatter-command) |
 | `soft-wrap`           | [editor.softwrap](./editor.md#editorsoft-wrap-section)
 | `text-width`          |  Maximum line length. Used for the `:reflow` command and soft-wrapping if `soft-wrap.wrap-at-text-width` is set, defaults to `editor.text-width`   |
 | `rulers`              | Overrides the `editor.rulers` config key for the language. |
@@ -101,6 +101,16 @@ with the following priorities:
 2. Extension: if there are no glob matches, any `file-types` string that matches
    the file extension of a given file wins. In the example above, the `"toml"`
    config matches files like `Cargo.toml` or `languages.toml`.
+
+### Configuring the formatter command
+
+[Command line expansions](./command-line.md#expansions) are supported in the arguments
+of the formatter command. In particular, the `%{buffer_name}` variable can be passed as
+argument to the formatter:
+
+```toml
+formatter = { command = "mylang-formatter" , args = ["--stdin", "--stdin-filename %{buffer_name}"] }
+```
 
 ## Language Server configuration
 

--- a/helix-core/src/command_line.rs
+++ b/helix-core/src/command_line.rs
@@ -357,13 +357,22 @@ pub struct Token<'a> {
     pub is_terminated: bool,
 }
 
-impl Token<'_> {
+impl<'a> Token<'a> {
     pub fn empty_at(content_start: usize) -> Self {
         Self {
             kind: TokenKind::Unquoted,
             content_start,
             content: Cow::Borrowed(""),
             is_terminated: false,
+        }
+    }
+
+    pub fn expand(content: impl Into<Cow<'a, str>>) -> Self {
+        Self {
+            kind: TokenKind::Expand,
+            content_start: 0,
+            content: content.into(),
+            is_terminated: true,
         }
     }
 }

--- a/helix-term/src/commands/typed.rs
+++ b/helix-term/src/commands/typed.rs
@@ -324,22 +324,20 @@ fn buffer_previous(
 fn write_impl(cx: &mut compositor::Context, path: Option<&str>, force: bool) -> anyhow::Result<()> {
     let config = cx.editor.config();
     let jobs = &mut cx.jobs;
-    {
-        let (view, doc) = current!(cx.editor);
+    let (view, doc) = current!(cx.editor);
 
-        if doc.trim_trailing_whitespace() {
-            trim_trailing_whitespace(doc, view.id);
-        }
-        if config.trim_final_newlines {
-            trim_final_newlines(doc, view.id);
-        }
-        if doc.insert_final_newline() {
-            insert_final_newline(doc, view.id);
-        }
-
-        // Save an undo checkpoint for any outstanding changes.
-        doc.append_changes_to_history(view);
+    if doc.trim_trailing_whitespace() {
+        trim_trailing_whitespace(doc, view.id);
     }
+    if config.trim_final_newlines {
+        trim_final_newlines(doc, view.id);
+    }
+    if doc.insert_final_newline() {
+        insert_final_newline(doc, view.id);
+    }
+
+    // Save an undo checkpoint for any outstanding changes.
+    doc.append_changes_to_history(view);
 
     let (view, doc) = current_ref!(cx.editor);
     let fmt = if config.auto_format {
@@ -738,26 +736,24 @@ pub fn write_all_impl(
         .collect();
 
     for (doc_id, target_view) in saves {
-        {
-            let doc = doc_mut!(cx.editor, &doc_id);
-            let view = view_mut!(cx.editor, target_view);
+        let doc = doc_mut!(cx.editor, &doc_id);
+        let view = view_mut!(cx.editor, target_view);
 
-            if doc.trim_trailing_whitespace() {
-                trim_trailing_whitespace(doc, target_view);
-            }
-            if config.trim_final_newlines {
-                trim_final_newlines(doc, target_view);
-            }
-            if doc.insert_final_newline() {
-                insert_final_newline(doc, target_view);
-            }
-
-            // Save an undo checkpoint for any outstanding changes.
-            doc.append_changes_to_history(view);
+        if doc.trim_trailing_whitespace() {
+            trim_trailing_whitespace(doc, target_view);
+        }
+        if config.trim_final_newlines {
+            trim_final_newlines(doc, target_view);
+        }
+        if doc.insert_final_newline() {
+            insert_final_newline(doc, target_view);
         }
 
-        let doc = doc!(cx.editor, &doc_id);
+        // Save an undo checkpoint for any outstanding changes.
+        doc.append_changes_to_history(view);
+
         let fmt = if options.auto_format && config.auto_format {
+            let doc = doc!(cx.editor, &doc_id);
             doc.auto_format(cx.editor).map(|fmt| {
                 let callback = make_format_callback(
                     doc_id,

--- a/helix-term/src/commands/typed.rs
+++ b/helix-term/src/commands/typed.rs
@@ -324,23 +324,26 @@ fn buffer_previous(
 fn write_impl(cx: &mut compositor::Context, path: Option<&str>, force: bool) -> anyhow::Result<()> {
     let config = cx.editor.config();
     let jobs = &mut cx.jobs;
-    let (view, doc) = current!(cx.editor);
+    {
+        let (view, doc) = current!(cx.editor);
 
-    if doc.trim_trailing_whitespace() {
-        trim_trailing_whitespace(doc, view.id);
-    }
-    if config.trim_final_newlines {
-        trim_final_newlines(doc, view.id);
-    }
-    if doc.insert_final_newline() {
-        insert_final_newline(doc, view.id);
+        if doc.trim_trailing_whitespace() {
+            trim_trailing_whitespace(doc, view.id);
+        }
+        if config.trim_final_newlines {
+            trim_final_newlines(doc, view.id);
+        }
+        if doc.insert_final_newline() {
+            insert_final_newline(doc, view.id);
+        }
+
+        // Save an undo checkpoint for any outstanding changes.
+        doc.append_changes_to_history(view);
     }
 
-    // Save an undo checkpoint for any outstanding changes.
-    doc.append_changes_to_history(view);
-
+    let (view, doc) = current_ref!(cx.editor);
     let fmt = if config.auto_format {
-        doc.auto_format().map(|fmt| {
+        doc.auto_format(cx.editor).map(|fmt| {
             let callback = make_format_callback(
                 doc.id(),
                 doc.version(),
@@ -483,8 +486,8 @@ fn format(cx: &mut compositor::Context, _args: Args, event: PromptEvent) -> anyh
         return Ok(());
     }
 
-    let (view, doc) = current!(cx.editor);
-    let format = doc.format().context(
+    let (view, doc) = current_ref!(cx.editor);
+    let format = doc.format(cx.editor).context(
         "A formatter isn't available, and no language server provides formatting capabilities",
     )?;
     let callback = make_format_callback(doc.id(), doc.version(), view.id, format, None);
@@ -735,24 +738,27 @@ pub fn write_all_impl(
         .collect();
 
     for (doc_id, target_view) in saves {
-        let doc = doc_mut!(cx.editor, &doc_id);
-        let view = view_mut!(cx.editor, target_view);
+        {
+            let doc = doc_mut!(cx.editor, &doc_id);
+            let view = view_mut!(cx.editor, target_view);
 
-        if doc.trim_trailing_whitespace() {
-            trim_trailing_whitespace(doc, target_view);
-        }
-        if config.trim_final_newlines {
-            trim_final_newlines(doc, target_view);
-        }
-        if doc.insert_final_newline() {
-            insert_final_newline(doc, target_view);
+            if doc.trim_trailing_whitespace() {
+                trim_trailing_whitespace(doc, target_view);
+            }
+            if config.trim_final_newlines {
+                trim_final_newlines(doc, target_view);
+            }
+            if doc.insert_final_newline() {
+                insert_final_newline(doc, target_view);
+            }
+
+            // Save an undo checkpoint for any outstanding changes.
+            doc.append_changes_to_history(view);
         }
 
-        // Save an undo checkpoint for any outstanding changes.
-        doc.append_changes_to_history(view);
-
+        let doc = doc!(cx.editor, &doc_id);
         let fmt = if options.auto_format && config.auto_format {
-            doc.auto_format().map(|fmt| {
+            doc.auto_format(cx.editor).map(|fmt| {
                 let callback = make_format_callback(
                     doc_id,
                     doc.version(),

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -5,6 +5,7 @@ use futures_util::future::BoxFuture;
 use futures_util::FutureExt;
 use helix_core::auto_pairs::AutoPairs;
 use helix_core::chars::char_is_word;
+use helix_core::command_line::{ExpansionKind, TokenKind, Tokenizer};
 use helix_core::diagnostic::DiagnosticProvider;
 use helix_core::doc_formatter::TextFormat;
 use helix_core::encoding::Encoding;
@@ -42,6 +43,7 @@ use helix_core::{
     ChangeSet, Diagnostic, LineEnding, Range, Rope, RopeBuilder, Selection, Syntax, Transaction,
 };
 
+use crate::expansion::Variable;
 use crate::{
     editor::Config,
     events::{DocumentDidChange, SelectionDidChange},
@@ -814,70 +816,120 @@ impl Document {
                 process.current_dir(doc_dir);
             }
 
-            let filename_in_args = fmt_args.iter().any(|arg| arg.contains("{}"));
-            let fmt_args = filename_in_args
-                .then_some(self.path().map(|path| {
-                    let path = path.to_string_lossy();
-                    Cow::Owned(
-                        fmt_args
-                            .iter()
-                            .map(|arg| arg.replace("{}", &path))
-                            .collect(),
-                    )
-                }))
-                .flatten()
-                .unwrap_or(Cow::Borrowed(fmt_args));
+            let fmt_args = fmt_args
+                .iter()
+                .map(|content| {
+                    let mut escaped = String::new();
+                    let mut start = 0;
 
-            process
-                .args(fmt_args.as_ref())
-                .stdin(Stdio::piped())
-                .stdout(Stdio::piped())
-                .stderr(Stdio::piped());
-
-            let formatting_future = async move {
-                let mut process = process
-                    .spawn()
-                    .map_err(|e| FormatterError::SpawningFailed {
-                        command: fmt_cmd.to_string_lossy().into(),
-                        error: e.kind(),
-                    })?;
-
-                let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
-                let input_text = text.clone();
-                let input_task = tokio::spawn(async move {
-                    to_writer(&mut stdin, (encoding::UTF_8, false), &input_text).await
-                    // Note that `stdin` is dropped here, causing the pipe to close. This can
-                    // avoid a deadlock with `wait_with_output` below if the process is waiting on
-                    // stdin to close before exiting.
-                });
-                let (input_result, output_result) = tokio::join! {
-                    input_task,
-                    process.wait_with_output(),
-                };
-                let _ = input_result.map_err(|_| FormatterError::BrokenStdin)?;
-                let output = output_result.map_err(|_| FormatterError::WaitForOutputFailed)?;
-
-                if !output.status.success() {
-                    if !output.stderr.is_empty() {
-                        let err = String::from_utf8_lossy(&output.stderr).to_string();
-                        log::error!("Formatter error: {}", err);
-                        return Err(FormatterError::NonZeroExitStatus(Some(err)));
+                    while let Some(offset) = content[start..].find('%') {
+                        let idx = start + offset;
+                        if content.as_bytes().get(idx + '%'.len_utf8()).copied() == Some(b'%') {
+                            // Treat two percents in a row as an escaped percent.
+                            escaped.push_str(&content[start..=idx]);
+                            // Skip over both percents.
+                            start = idx + ('%'.len_utf8() * 2);
+                        } else {
+                            // Otherwise interpret the percent as an expansion. Push up to (but not
+                            // including) the percent token.
+                            escaped.push_str(&content[start..idx]);
+                            // Then parse the expansion,
+                            let mut tokenizer = Tokenizer::new(&content[idx..], true);
+                            let token = tokenizer
+                                .parse_percent_token()
+                                .unwrap()
+                                .map_err(|err| anyhow!("{err}"))?;
+                            if matches!(token.kind, TokenKind::Expansion(ExpansionKind::Variable)) {
+                                let var = Variable::from_name(&token.content).ok_or_else(|| {
+                                    anyhow!("unknown variable '{}'", token.content)
+                                })?;
+                                if matches!(var, Variable::BufferName) {
+                                    let expanded = if let Some(path) = self.relative_path() {
+                                        Cow::Owned(path.to_string_lossy().into_owned())
+                                    } else {
+                                        Cow::Borrowed(crate::document::SCRATCH_BUFFER_NAME)
+                                    };
+                                    escaped.push_str(expanded.as_ref());
+                                } else {
+                                    bail!(
+                                        "unexpected variable in format command '{}'",
+                                        var.as_str()
+                                    );
+                                }
+                                // and move forward to the end of the expansion.
+                                start = idx + tokenizer.pos();
+                            } else {
+                                bail!("unexpected token in format command '{}'", token.content);
+                            }
+                        }
                     }
 
-                    return Err(FormatterError::NonZeroExitStatus(None));
-                } else if !output.stderr.is_empty() {
-                    log::debug!(
-                        "Formatter printed to stderr: {}",
-                        String::from_utf8_lossy(&output.stderr)
-                    );
+                    if escaped.is_empty() {
+                        Ok(Cow::Borrowed(content))
+                    } else {
+                        escaped.push_str(&content[start..]);
+                        Ok(Cow::Owned(escaped))
+                    }
+                })
+                .collect::<anyhow::Result<Vec<_>>>();
+
+            match fmt_args {
+                Ok(fmt_args) => {
+                    process
+                        .args(fmt_args.iter().map(Cow::as_ref))
+                        .stdin(Stdio::piped())
+                        .stdout(Stdio::piped())
+                        .stderr(Stdio::piped());
+
+                    let formatting_future = async move {
+                        let mut process =
+                            process
+                                .spawn()
+                                .map_err(|e| FormatterError::SpawningFailed {
+                                    command: fmt_cmd.to_string_lossy().into(),
+                                    error: e.kind(),
+                                })?;
+
+                        let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
+                        let input_text = text.clone();
+                        let input_task = tokio::spawn(async move {
+                            to_writer(&mut stdin, (encoding::UTF_8, false), &input_text).await
+                            // Note that `stdin` is dropped here, causing the pipe to close. This can
+                            // avoid a deadlock with `wait_with_output` below if the process is waiting on
+                            // stdin to close before exiting.
+                        });
+                        let (input_result, output_result) = tokio::join! {
+                            input_task,
+                            process.wait_with_output(),
+                        };
+                        let _ = input_result.map_err(|_| FormatterError::BrokenStdin)?;
+                        let output =
+                            output_result.map_err(|_| FormatterError::WaitForOutputFailed)?;
+
+                        if !output.status.success() {
+                            if !output.stderr.is_empty() {
+                                let err = String::from_utf8_lossy(&output.stderr).to_string();
+                                log::error!("Formatter error: {}", err);
+                                return Err(FormatterError::NonZeroExitStatus(Some(err)));
+                            }
+
+                            return Err(FormatterError::NonZeroExitStatus(None));
+                        } else if !output.stderr.is_empty() {
+                            log::debug!(
+                                "Formatter printed to stderr: {}",
+                                String::from_utf8_lossy(&output.stderr)
+                            );
+                        }
+
+                        let str = std::str::from_utf8(&output.stdout)
+                            .map_err(|_| FormatterError::InvalidUtf8Output)?;
+
+                        Ok(helix_core::diff::compare_ropes(&text, &Rope::from(str)))
+                    };
+                    return Some(formatting_future.boxed());
                 }
-
-                let str = std::str::from_utf8(&output.stdout)
-                    .map_err(|_| FormatterError::InvalidUtf8Output)?;
-
-                Ok(helix_core::diff::compare_ropes(&text, &Rope::from(str)))
-            };
-            return Some(formatting_future.boxed());
+                Err(e) => log::error!("{e}"),
+            }
         };
 
         let text = self.text.clone();

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -5,7 +5,8 @@ use futures_util::future::BoxFuture;
 use futures_util::FutureExt;
 use helix_core::auto_pairs::AutoPairs;
 use helix_core::chars::char_is_word;
-use helix_core::command_line::{ExpansionKind, TokenKind, Tokenizer};
+
+use helix_core::command_line::Token;
 use helix_core::diagnostic::DiagnosticProvider;
 use helix_core::doc_formatter::TextFormat;
 use helix_core::encoding::Encoding;
@@ -43,7 +44,7 @@ use helix_core::{
     ChangeSet, Diagnostic, LineEnding, Range, Rope, RopeBuilder, Selection, Syntax, Transaction,
 };
 
-use crate::expansion::Variable;
+use crate::expansion::expand;
 use crate::{
     editor::Config,
     events::{DocumentDidChange, SelectionDidChange},
@@ -779,9 +780,12 @@ impl Document {
 
     /// The same as [`format`], but only returns formatting changes if auto-formatting
     /// is configured.
-    pub fn auto_format(&self) -> Option<BoxFuture<'static, Result<Transaction, FormatterError>>> {
+    pub fn auto_format(
+        &self,
+        editor: &Editor,
+    ) -> Option<BoxFuture<'static, Result<Transaction, FormatterError>>> {
         if self.language_config()?.auto_format {
-            self.format()
+            self.format(editor)
         } else {
             None
         }
@@ -791,7 +795,10 @@ impl Document {
     /// to format it nicely.
     // We can't use anyhow::Result here since the output of the future has to be
     // clonable to be used as shared future. So use a custom error type.
-    pub fn format(&self) -> Option<BoxFuture<'static, Result<Transaction, FormatterError>>> {
+    pub fn format(
+        &self,
+        editor: &Editor,
+    ) -> Option<BoxFuture<'static, Result<Transaction, FormatterError>>> {
         if let Some((fmt_cmd, fmt_args)) = self
             .language_config()
             .and_then(|c| c.formatter.as_ref())
@@ -818,59 +825,7 @@ impl Document {
 
             let fmt_args = fmt_args
                 .iter()
-                .map(|content| {
-                    let mut escaped = String::new();
-                    let mut start = 0;
-
-                    while let Some(offset) = content[start..].find('%') {
-                        let idx = start + offset;
-                        if content.as_bytes().get(idx + '%'.len_utf8()).copied() == Some(b'%') {
-                            // Treat two percents in a row as an escaped percent.
-                            escaped.push_str(&content[start..=idx]);
-                            // Skip over both percents.
-                            start = idx + ('%'.len_utf8() * 2);
-                        } else {
-                            // Otherwise interpret the percent as an expansion. Push up to (but not
-                            // including) the percent token.
-                            escaped.push_str(&content[start..idx]);
-                            // Then parse the expansion,
-                            let mut tokenizer = Tokenizer::new(&content[idx..], true);
-                            let token = tokenizer
-                                .parse_percent_token()
-                                .unwrap()
-                                .map_err(|err| anyhow!("{err}"))?;
-                            if matches!(token.kind, TokenKind::Expansion(ExpansionKind::Variable)) {
-                                let var = Variable::from_name(&token.content).ok_or_else(|| {
-                                    anyhow!("unknown variable '{}'", token.content)
-                                })?;
-                                if matches!(var, Variable::BufferName) {
-                                    let expanded = if let Some(path) = self.relative_path() {
-                                        Cow::Owned(path.to_string_lossy().into_owned())
-                                    } else {
-                                        Cow::Borrowed(crate::document::SCRATCH_BUFFER_NAME)
-                                    };
-                                    escaped.push_str(expanded.as_ref());
-                                } else {
-                                    bail!(
-                                        "unexpected variable in format command '{}'",
-                                        var.as_str()
-                                    );
-                                }
-                                // and move forward to the end of the expansion.
-                                start = idx + tokenizer.pos();
-                            } else {
-                                bail!("unexpected token in format command '{}'", token.content);
-                            }
-                        }
-                    }
-
-                    if escaped.is_empty() {
-                        Ok(Cow::Borrowed(content))
-                    } else {
-                        escaped.push_str(&content[start..]);
-                        Ok(Cow::Owned(escaped))
-                    }
-                })
+                .map(|content| expand(editor, Token::expand(content)))
                 .collect::<anyhow::Result<Vec<_>>>();
 
             match fmt_args {

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -5,7 +5,6 @@ use futures_util::future::BoxFuture;
 use futures_util::FutureExt;
 use helix_core::auto_pairs::AutoPairs;
 use helix_core::chars::char_is_word;
-
 use helix_core::command_line::Token;
 use helix_core::diagnostic::DiagnosticProvider;
 use helix_core::doc_formatter::TextFormat;
@@ -44,10 +43,10 @@ use helix_core::{
     ChangeSet, Diagnostic, LineEnding, Range, Rope, RopeBuilder, Selection, Syntax, Transaction,
 };
 
-use crate::expansion::expand;
 use crate::{
     editor::Config,
     events::{DocumentDidChange, SelectionDidChange},
+    expansion,
     view::ViewPosition,
     DocumentId, Editor, Theme, View, ViewId,
 };
@@ -823,68 +822,68 @@ impl Document {
                 process.current_dir(doc_dir);
             }
 
-            let fmt_args = fmt_args
+            let args = match fmt_args
                 .iter()
-                .map(|content| expand(editor, Token::expand(content)))
-                .collect::<anyhow::Result<Vec<_>>>();
-
-            match fmt_args {
-                Ok(fmt_args) => {
-                    process
-                        .args(fmt_args.iter().map(Cow::as_ref))
-                        .stdin(Stdio::piped())
-                        .stdout(Stdio::piped())
-                        .stderr(Stdio::piped());
-
-                    let formatting_future = async move {
-                        let mut process =
-                            process
-                                .spawn()
-                                .map_err(|e| FormatterError::SpawningFailed {
-                                    command: fmt_cmd.to_string_lossy().into(),
-                                    error: e.kind(),
-                                })?;
-
-                        let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
-                        let input_text = text.clone();
-                        let input_task = tokio::spawn(async move {
-                            to_writer(&mut stdin, (encoding::UTF_8, false), &input_text).await
-                            // Note that `stdin` is dropped here, causing the pipe to close. This can
-                            // avoid a deadlock with `wait_with_output` below if the process is waiting on
-                            // stdin to close before exiting.
-                        });
-                        let (input_result, output_result) = tokio::join! {
-                            input_task,
-                            process.wait_with_output(),
-                        };
-                        let _ = input_result.map_err(|_| FormatterError::BrokenStdin)?;
-                        let output =
-                            output_result.map_err(|_| FormatterError::WaitForOutputFailed)?;
-
-                        if !output.status.success() {
-                            if !output.stderr.is_empty() {
-                                let err = String::from_utf8_lossy(&output.stderr).to_string();
-                                log::error!("Formatter error: {}", err);
-                                return Err(FormatterError::NonZeroExitStatus(Some(err)));
-                            }
-
-                            return Err(FormatterError::NonZeroExitStatus(None));
-                        } else if !output.stderr.is_empty() {
-                            log::debug!(
-                                "Formatter printed to stderr: {}",
-                                String::from_utf8_lossy(&output.stderr)
-                            );
-                        }
-
-                        let str = std::str::from_utf8(&output.stdout)
-                            .map_err(|_| FormatterError::InvalidUtf8Output)?;
-
-                        Ok(helix_core::diff::compare_ropes(&text, &Rope::from(str)))
-                    };
-                    return Some(formatting_future.boxed());
+                .map(|content| expansion::expand(editor, Token::expand(content)))
+                .collect::<Result<Vec<_>, _>>()
+            {
+                Ok(args) => args,
+                Err(err) => {
+                    log::error!("Failed to expand formatter arguments: {err}");
+                    return None;
                 }
-                Err(e) => log::error!("{e}"),
-            }
+            };
+
+            process
+                .args(args.iter().map(AsRef::as_ref))
+                .stdin(Stdio::piped())
+                .stdout(Stdio::piped())
+                .stderr(Stdio::piped());
+
+            let formatting_future = async move {
+                let mut process = process
+                    .spawn()
+                    .map_err(|e| FormatterError::SpawningFailed {
+                        command: fmt_cmd.to_string_lossy().into(),
+                        error: e.kind(),
+                    })?;
+
+                let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
+                let input_text = text.clone();
+                let input_task = tokio::spawn(async move {
+                    to_writer(&mut stdin, (encoding::UTF_8, false), &input_text).await
+                    // Note that `stdin` is dropped here, causing the pipe to close. This can
+                    // avoid a deadlock with `wait_with_output` below if the process is waiting on
+                    // stdin to close before exiting.
+                });
+                let (input_result, output_result) = tokio::join! {
+                    input_task,
+                    process.wait_with_output(),
+                };
+                let _ = input_result.map_err(|_| FormatterError::BrokenStdin)?;
+                let output = output_result.map_err(|_| FormatterError::WaitForOutputFailed)?;
+
+                if !output.status.success() {
+                    if !output.stderr.is_empty() {
+                        let err = String::from_utf8_lossy(&output.stderr).to_string();
+                        log::error!("Formatter error: {}", err);
+                        return Err(FormatterError::NonZeroExitStatus(Some(err)));
+                    }
+
+                    return Err(FormatterError::NonZeroExitStatus(None));
+                } else if !output.stderr.is_empty() {
+                    log::debug!(
+                        "Formatter printed to stderr: {}",
+                        String::from_utf8_lossy(&output.stderr)
+                    );
+                }
+
+                let str = std::str::from_utf8(&output.stdout)
+                    .map_err(|_| FormatterError::InvalidUtf8Output)?;
+
+                Ok(helix_core::diff::compare_ropes(&text, &Rope::from(str)))
+            };
+            return Some(formatting_future.boxed());
         };
 
         let text = self.text.clone();

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -814,8 +814,22 @@ impl Document {
                 process.current_dir(doc_dir);
             }
 
+            let filename_in_args = fmt_args.iter().any(|arg| arg.contains("{}"));
+            let fmt_args = filename_in_args
+                .then_some(self.path().map(|path| {
+                    let path = path.to_string_lossy();
+                    Cow::Owned(
+                        fmt_args
+                            .iter()
+                            .map(|arg| arg.replace("{}", &path))
+                            .collect(),
+                    )
+                }))
+                .flatten()
+                .unwrap_or(Cow::Borrowed(fmt_args));
+
             process
-                .args(fmt_args)
+                .args(fmt_args.as_ref())
                 .stdin(Stdio::piped())
                 .stdout(Stdio::piped())
                 .stderr(Stdio::piped());

